File: app/test/domain/plan_schema_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:gym_coach/domain/plan_schema.dart';

void main() {
  group('looksLikePlanJson', () {
    test('returns true for pure JSON block', () {
      expect(looksLikePlanJson('{"weeks":12}'), isTrue);
    });
    test('returns true for JSON with leading/trailing whitespace', () {
      expect(looksLikePlanJson('  \n  {"weeks":12}\n  '), isTrue);
    });
    test('returns false for chat text', () {
      expect(looksLikePlanJson('好的，我已经为你准备好计划'), isFalse);
    });
  });

  group('validatePlanJson', () {
    test('accepts a complete valid plan', () {
      final raw = '''
      {
        "weeks": 12,
        "weekly_structure": "4 天训练 + 3 天休息",
        "goal_summary": "增肌",
        "training_days": [],
        "daily_meals": [],
        "startDate": "2026-08-14T00:00:00.000Z"
      }
      ''';
      final result = validatePlanJson(raw);
      expect(result.error, isNull);
      expect(result.plan, isNotNull);
      expect(result.plan!.weeks, 12);
    });

    test('rejects plan missing weeks field', () {
      final raw = '{"weekly_structure":"x","goal_summary":"y","training_days":[],"daily_meals":[],"startDate":"2026-08-14T00:00:00.000Z"}';
      final result = validatePlanJson(raw);
      expect(result.plan, isNull);
      expect(result.error, contains('weeks'));
    });

    test('rejects malformed JSON', () {
      final result = validatePlanJson('{not json}');
      expect(result.plan, isNull);
      expect(result.error, isNotNull);
    });
  });
}

File: app/android/app/src/test/java/com/gymcoach/gym_coach/domain/PlanSchemaTest.kt
package com.gymcoach.gym_coach.domain

import com.gymcoach.gym_coach.domain.plan.PlanSchema
import com.gymcoach.gym_coach.domain.plan.PlanValidationResult
import com.google.common.truth.Truth.assertThat
import org.junit.Test

class PlanSchemaTest {

    @Test
    fun `looksLikePlanJson returns true for valid json object`() {
        assertThat(PlanSchema.looksLikePlanJson("""{"weeks": 1}""")).isTrue()
    }

    @Test
    fun `looksLikePlanJson returns false for plain text`() {
        assertThat(PlanSchema.looksLikePlanJson("好的，我问几个问题")).isFalse()
    }

    @Test
    fun `looksLikePlanJson returns false for non-json text`() {
        assertThat(PlanSchema.looksLikePlanJson("hi")).isFalse()
    }

    @Test
    fun `validatePlanJson accepts complete json`() {
        val json = """
            {
              "weeks": 4,
              "weekly_structure": "3 lift + 1 active recovery",
              "goal_summary": "Build muscle",
              "training_days": [],
              "daily_meals": []
            }
        """.trimIndent()
        val result = PlanSchema.validatePlanJson(json)
        assertThat(result).isInstanceOf(PlanValidationResult.Success::class.java)
        assertThat((result as PlanValidationResult.Success).plan.weeks).isEqualTo(4)
    }

    @Test
    fun `validatePlanJson rejects when weeks missing`() {
        val json = """
            {
              "weekly_structure": "x",
              "goal_summary": "y",
              "training_days": [],
              "daily_meals": []
            }
        """.trimIndent()
        val result = PlanSchema.validatePlanJson(json)
        assertThat(result).isInstanceOf(PlanValidationResult.Failure::class.java)
    }

    @Test
    fun `validatePlanJson rejects malformed json`() {
        val result = PlanSchema.validatePlanJson("{not json}")
        assertThat(result).isInstanceOf(PlanValidationResult.Failure::class.java)
    }
}
